kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/SFTPClient.scala

```diff
@@ -0,0 +1,249 @@
+    } yield ftpFiles)
+      .recover {
+        case e: Exception =>
+          logger.error(s"SFTPClient Error obtaining resources from pathname $pathname. Caused by ${ExceptionUtils.getStackTrace(e)}")
+          List()
+      }.get
+  }
+
+  private def connectChannel(channel: ChannelSftp): Unit = {
+    maybeDataTimeout
+      .fold(channel.connect())(dataTimeout => channel.connect(dataTimeout))
+  }
+
+  private def fetchFiles(pathname: String, channel: ChannelSftp): Try[List[FTPFile]] = {
+    Try {
+      channel.ls(pathname)
+        .asScala
+        .toList
+        .map(file => file.asInstanceOf[ChannelSftp#LsEntry])
+        .filter(lsEntry => lsEntry.getFilename != "." && lsEntry.getFilename != "..")
+        .map(lsEntry => createFtpFile(lsEntry))
+    }
+  }
+
+  private def createFtpFile(lsEntry: ChannelSftp#LsEntry) = {
+    val ftpFile: FTPFile = new FTPFile()
+    ftpFile.setType(0)
+    ftpFile.setName(lsEntry.getFilename)
+    ftpFile.setSize(lsEntry.getAttrs.getSize)
+
+    val dateFormat = new SimpleDateFormat("EEE MMM d HH:mm:ss zzz uuuu")
+    val calendar = Calendar.getInstance()
+    calendar.setTime(dateFormat.parse(lsEntry.getAttrs.getMtimeString))
+    ftpFile.setTimestamp(calendar)
+    ftpFile
+  }
+
+  private def getSessionAndChannel(username: Username,
+                                   password: Password): Try[(Session, ChannelSftp)] = {
+    for {
+      session <- openSession(username, password)
+      channel <- createChannel(session)
+    } yield (session, channel)
+  }
+
+  /**
+    * Open a channel with protocol [sftp].
+    */
+  private val createChannel: Session => Try[ChannelSftp] = {
+    session => Try(session.openChannel("sftp").asInstanceOf[ChannelSftp])
+  }
+
+  /**
+    * Create and open a session in default port 22 or in a specific one using hostname, username and password
+    */
+  private val openSession: (Username, Password) => Try[Session] = {
+    (username, password) =>
+      Try {
+        val jsch = new JSch()
+        val hostname = getHostname(username)
+        val session = createSession(username, jsch, hostname)
+        session.setPassword(password.value)
+        setSessionConfig(session)
+        connectSession(session)
+        session
+      }
+  }
+
+  private def setSessionConfig(session: Session): Unit = {
+    val config = new Properties();
+    config.put("StrictHostKeyChecking", "no")
+    session.setConfig(config);
+  }
+
+  private def connectSession(session: Session): Unit = {
+    maybeConnectTimeout
+      .fold(session.connect()) {
+        connectTimeout => session.connect(connectTimeout)
+      }
+  }
+
+  private def createSession(username: Username, jsch: JSch, hostname: String): Session = {
+    maybeExplicitPort
+      .fold(jsch.getSession(username.value, hostname)) {
+        explicitPort => jsch.getSession(username.value, hostname, explicitPort)
+      }
+  }
+
+  private def getHostname(username: Username): String = {
```

Comment:
It would be good to return an Either instead of throwing an exception here.

Reply:
If you don't mind I will use a Try to then being able to have a for comprehension in openSession function

Reply:
That'd be fine :+1: